decision-records/006-javascript-compatibility.md

```diff
@@ -0,0 +1,161 @@
+
+### Tooling isn't comprehensive
+
+We're not sure of the exact coverage our tools will provide for either checking the use of features or transforming them down to ES 2015.
+
+For example, [eslint-plugin-es-x](https://eslint-community.github.io/eslint-plugin-es-x/) does not cover Web APIs and there's an engineering cost to [using the transpilation step for detecting whether polyfills would need including](https://github.com/babel/babel-polyfills/pull/13/files) reliably (especially if we guard the execution of some of our features to the presence of specific APIs in the browsers, for which we wouldn't want a polyfill).
+
+This means we still need some level of manual checks, which are susceptible to human errors.
+
+### Tooling evolution
+
+The tools we plan to use may evolve to only support a more modern baseline than ES 2015 (similarly to how Rollup stopped supporting IE8 at some point). This could force us to pin our tools to a specific version if we want to keep benefiting from them, devise a new approach for ensuring our compatibility, or revisit our baseline (this would require a major release though).
+
+### Shipped code monitoring
+
+We're doing heavier transformations of our code between what we author and what we ship. This provides more room for malicious code to potentially be introduced by our tooling (supply chain attacks).
+
+Monitoring the changes from one release to another may also be trickier as tooling updates may introduce extra changes to the code (for example refactoring or reordering...) that would make comparison harder.
+
+## Alternatives considered
+
+### Adding the styling hook using type="module"
+
+The inline script adding the `govuk-frontend-supported` class cannot be run in a `<script type="module">`. This would defer the execution of the script adding the class until after the page loads. This defeats the purpose of that script, as it needs to run immediately so it prevents flashes of non-interactive styles.
+
+### Shipping the code we author
+
+We could author ES2015 code and ship what we author.
+
+While we could lint away some compatibility concerns, this would require more manual testing to ensure we match what the browsers we support can parse and execute.
+
+Due to the more manual nature, it would also be more prone to bugs and regressions.
+
+And finally, due to the extra knowledge necessary to know what's compatible or not, it would require extra effort from developers and contributors to offer changes.
+
+### Guarding based on newer ES versions and letting JavaScript error
+
+Instead of using the availability of `HTMLScriptElement.prototype.noModule` to set the `govuk-frontend-supported` class (which governs both styling and execution of our components), we could:
+
+- pick a more modern feature to check (for example `Promise.prototype.finally`)
+- transpile down to the more modern set of browsers that support it (rather than all the way down to 2017)
```

Comment:
Shall we stick to ES2015 rather than 2017? (The year that browsers rolled out support)

```suggestion
- transpile down to the more modern set of browsers that support it (rather than all the way down to ES2015)
```